example/example20.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+  example/example20.cpp -- Usage of structured numpy dtypes
+
+  Copyright (c) 2016 Ivan Smirnov
+
+  All rights reserved. Use of this source code is governed by a
+  BSD-style license that can be found in the LICENSE file.
+*/
+
+#include "example.h"
+
+#include <pybind11/numpy.h>
+#include <cstdint>
+#include <iostream>
+
+namespace py = pybind11;
+
+struct SimpleStruct {
+    bool x;
+    uint32_t y;
+    float z;
+};
+
+std::ostream& operator<<(std::ostream& os, const SimpleStruct& v) {
+    return os << "s:" << v.x << "," << v.y << "," << v.z;
+}
+
+struct PackedStruct {
+    bool x;
+    uint32_t y;
+    float z;
+} __attribute__((packed));
+
+std::ostream& operator<<(std::ostream& os, const PackedStruct& v) {
+    return os << "p:" << v.x << "," << v.y << "," << v.z;
+}
+
+struct NestedStruct {
+    SimpleStruct a;
+    PackedStruct b;
+} __attribute__((packed));
+
+std::ostream& operator<<(std::ostream& os, const NestedStruct& v) {
+    return os << "n:a=" << v.a << ";b=" << v.b;
+}
+
+struct PartialStruct {
+    bool x;
+    uint32_t y;
+    float z;
+    long dummy2;
+};
+
+struct PartialNestedStruct {
+    long dummy1;
+    PartialStruct a;
+    long dummy2;
+};
+
+struct UnboundStruct { };
+
+template <typename T>
+py::array mkarray_via_buffer(size_t n) {
+    return py::array(py::buffer_info(nullptr, sizeof(T),
+                                     py::format_descriptor<T>::format(),
+                                     1, { n }, { sizeof(T) }));
+}
+
+template <typename S>
+py::array_t<S, 0> create_recarray(size_t n) {
+    auto arr = mkarray_via_buffer<S>(n);
+    auto ptr = static_cast<S*>(arr.request().ptr);
```

Comment:
Better to save the return value of request() in a local variable. Otherwise the buffer is released before you fill it below. No functional difference in most cases, but I'd say it's better practice to be careful.